### ISSUE TITLE
concat() fails to recognize identical SPWs despite large fqtol

### ISSUE THREAD

**Issue by jwagnerhki** [casacore/casacore#1336]
Under at least CASA 6.6, if not earlier, the 'fqtol' parameter of concat() (invoked by importfitsidi()) does not seem to work as documented.

I have two 32-IF FITS-IDI files with identical sky frequencies. They differ only in FQ table REF_FREQ=228162796875 for one and REF_FREQ=228221390625 for the other. After individual import of each FITS-IDI file using importfitsidi , browsetable of the SPECTRAL_WINDOW data shows that both MS have the same 32 frequency entries (e.g. 2.2828e+11 exists in both MS sets).

Nevertheless, concatenation of the two MS results in a new MS with 64 frequency entries, 32 of which are duplicates. This happens even when using a very large freqtol='30MHz'. The same duplication of SPWs happens also when importing the FITS-IDI not individually but together, i.e., importfitsidi(vis=['testfits-part1.fitsidi','testfits-part2.fitsidi']). 

I haven't been able to trace the issue, I only suspect that some check around [MSConcat.cc#L3590](https://github.com/casacore/casacore/blob/master/ms/MSOper/MSConcat.cc#L3590) is perhaps not working as intended, and fails to  recognize a "local" and the "other" SPW as describing an identical sky frequency, bandwidth, and nr of channels.

The two sample FITS-IDI (~240 kB each) are at:

https://www.dropbox.com/scl/fi/45q8it51joxe3radb5pz9/testfits-part1.fitsidi?rlkey=al8gmpdbls39iffhj3314xo99&dl=0
https://www.dropbox.com/scl/fi/45q8it51joxe3radb5pz9/testfits-part1.fitsidi?rlkey=al8gmpdbls39iffhj3314xo99&dl=0

[testfitsload.log](https://github.com/casacore/casacore/files/14190900/testfitsload.log)
[testfits-part1-casaSPW.txt](https://github.com/casacore/casacore/files/14190901/testfits-part1-casaSPW.txt)
[testfits-part2-casaSPW.txt](https://github.com/casacore/casacore/files/14190902/testfits-part2-casaSPW.txt)
[testfits-parts-FQ.txt](https://github.com/casacore/casacore/files/14190905/testfits-parts-FQ.txt)
[test_1-casaSPW.txt](https://github.com/casacore/casacore/files/14190959/test_1-casaSPW.txt)
[test_2-casaSPW.txt](https://github.com/casacore/casacore/files/14190960/test_2-casaSPW.txt)
[printFitsFQ.py.txt](https://github.com/casacore/casacore/files/14190969/printFitsFQ.py.txt)
[testfitsload.py.txt](https://github.com/casacore/casacore/files/14190970/testfitsload.py.txt)

**Comment by dpetry** [comment]
Hi, I am one of the authors of concat. Could you provide the listobs output of the two input MSs for your concat call?

**Comment by jwagnerhki** [comment]
Hi, thanks for looking into this. The listobs outputs for the two MS sets are below. 

The sky frequencies in both are the same, except that frequencies from part_1.ms 228191.7891 MHz (spw#0) 228250.3828 MHz (spw#1) till 230008.1953 (spw#31) appear in part_2.ms in a "rotated-by-one" fashion, namely, part_2.ms starts with 228250.3828 MHz (spw#0) and ends with 230008.1953 (spw#30) 228191.7891 MHz (spw#31). 

[part_1.listobs.log](https://github.com/casacore/casacore/files/14293056/part_1.listobs.log)
[part_1.listobs.txt](https://github.com/casacore/casacore/files/14293058/part_1.listobs.txt)
[part_2.listobs.log](https://github.com/casacore/casacore/files/14293059/part_2.listobs.log)
[part_2.listobs.txt](https://github.com/casacore/casacore/files/14293060/part_2.listobs.txt)

**Comment by jwagnerhki** [comment]
Unassigned myself - cannot create a pull request since I do not quite follow the concat code, and do not have a complete build and CASA test environment.

**Comment by dpetry** [comment]
Hi, sorry, didn't have time to look at this, yet. Does your last message mean that you stopped working on this?

**Comment by jwagnerhki** [comment]
Alas yes, I gave up trying to set up a working casa + casacore build and debug environment.

The MS concat issue with SPWs persists, though, and affects a few PIs of EHT VLBI projects.

It would be great if you or another casacore developer could reproduce the issue using the datasets and logs provided here, and ideally fix the MS concatenation - assuming that the duplication of SPWs is indeed a bug and not a feature.

**Comment by dpetry** [comment]
OK, I will try to have a look ...

**Comment by dpetry** [comment]
Did you try to import the two FITS IDI files separately into MSs and then concat them?
If so, could you attach the log output of concat for that?

**Comment by jwagnerhki** [comment]
Yes, tried that approach, too, see the script testfitsload.py.txt linked above, or this copy of code:

```
files = ['testfits-part1.fitsidi', 'testfits-part2.fitsidi']
parts = ['part_%d.ms' % (n+1) for n in range(len(files))]
for n in range(len(files)):
    importfitsidi(fitsidifile=files[n], vis=parts[n], constobsid=True, scanreindexgap_s=15)
concat(vis=parts, concatvis='test_2.ms', freqtol='30MHz')
listobs('test_2.ms',verbose=True,listfile='test_2.listobs')
```

The output on the CASA ipython console of the above is [testfits_2-console-log.txt](https://github.com/casacore/casacore/files/14404791/testfits_2-console-log.txt), and the CASA log file is 
[testfits_2-casalog.txt](https://github.com/casacore/casacore/files/14404800/testfits_2-casalog.txt)

The concat part of the CASA log is

```
...
INFO	concat::::casa	copying part_1.ms to test_2.ms
INFO	concat::::casa	concatenating part_2.ms into test_2.ms
INFO	MSConcat::concatenate 	Appending /data/eht2018/e18g27/band4/fits/part_2.ms to /data/eht2018/e18g27/band4/fits/test_2.ms
INFO	MSConcat::concatenate 	No valid state tables present. Result won't have one either.
INFO	MSConcat::concatenate 	Added 32 rows and matched 0 from the data description subtable
INFO	MSConcat::concatenate 	Added 32 rows and matched 0 from the spectral window subtable
...
```

so basically concat fails to match the shifted-by-1 but otherwise (in terms of sky freq, bandwidth) identical SPWs.

You can find the two small FITS IDI example files linked in the original post.

**Comment by dpetry** [comment]
@jwagnerhki , I tried to download the two FITS IDI files but you posted the wrong link for the second file.
You posted the link for the first one twice.

**Comment by dpetry** [comment]
@jwagnerhki , the difference between the SPECTRAL_WINDOW tables in the two imported MSs
was in the IF_CONV_CHAIN column. This can be important in principle, but should probably be
ignored here. So you can just set all cells in that column to 1 (integer one) as demonstrated
in the following little CASA script

```
import numpy as np
tb.open('testfits-part1.ms/SPECTRAL_WINDOW', nomodify=False)
a = tb.getcol('IF_CONV_CHAIN')
b = np.ones(len(a), dtype='int')
tb.putcol('IF_CONV_CHAIN', b)
tb.close()
tb.open('testfits-part2.ms/SPECTRAL_WINDOW', nomodify=False)
tb.putcol('IF_CONV_CHAIN', b)
tb.close()
```
After that, the concat nicely merges the two MSs with only 32 SPWs
in the output MS.
